crates/astria-sequencer-relayer/src/data_availability.rs

```diff
@@ -352,128 +309,143 @@ impl CelestiaClient {
+            .collect();
+        // retain rollup datas whose block hash matches the block hash of the namespaced data
+        rollup_datas
+            .retain(|_, rollup_data| namespace_data.data.block_hash == rollup_data.data.block_hash);
+
+        // retain rollup datas that can be verified
+        rollup_datas.retain(|namespace, rollup_data| {
+            if let Err(e) = rollup_data
+                .data
+                .hash_json_serialized_bytes()
+                .wrap_err("failed hashing json serialized rollup namespace data")
+                .and_then(|hash| {
+                    Signature::try_from(&*rollup_data.signature)
+                        .map(|signature| (hash, signature))
+                        .wrap_err(
+                            "failed constructing signature from signature bytes of namespace data",
+                        )
+                })
+                .and_then(|(hash, signature)| {
+                    verification_key.verify(&signature, &hash).wrap_err(
+                        "applying verification key to signature and hash generated from rollup \
+                         namespace data failed",
+                    )
+                })
+            {
+                warn!(error.msg = %e, error.cause = ?e, %namespace, "failed verifying rollup namespace data");
+                return false;
+            }
+            true
+        });
+
+        // finally, extract the rollup txs from the rollup datas
+        let rollup_txs = rollup_datas
+            .into_iter()
+            .map(|(namespace, rollup_datas)| (namespace, rollup_datas.data.rollup_txs))
+            .collect();
+        Ok(SequencerBlockData {
+            block_hash: namespace_data.data.block_hash.clone(),
+            header: namespace_data.data.header.clone(),
+            last_commit: namespace_data.data.last_commit.clone(),
+            rollup_txs,
+        })
+    }
+}
 
-        let Some(rollup_data_for_block) = rollup_datas.next() else {
-            return Ok(None);
+fn assemble_blobs_from_sequencer_block_data(
+    block_data: SequencerBlockData,
+    signing_key: &SigningKey,
+    verification_key: VerificationKey,
+) -> eyre::Result<Vec<blob::Blob>> {
+    let mut blobs = Vec::with_capacity(block_data.rollup_txs.len() + 1);
+    let mut namespaces = Vec::with_capacity(block_data.rollup_txs.len() + 1);
+    for (namespace, txs) in block_data.rollup_txs {
+        let rollup_namespace_data = RollupNamespaceData {
+            block_hash: block_data.block_hash.clone(),
+            rollup_txs: txs,
         };
-
-        // there should NOT be multiple datas with the same block hash and signer
-        if rollup_datas.next().is_some() {
-            bail!("multiple rollup datas with the same block hash and signer");
-        }
-
-        Ok(Some(rollup_data_for_block.data.rollup_txs))
+        let data = rollup_namespace_data
+            .to_signed(signing_key, verification_key)
```

Comment:
maybe out of scope, but why does this need the verification key? seems unnecessary (it can be derived from the signing key) is it more so we don't have to re-derive it?

Reply:
Makes sense to change it. But I think I was just calling the provided method without thinking about it.

Make a follow-up issue?

Reply:
It's even worse then that - the verification key is literally part of the signingkey here: https://docs.rs/ed25519-consensus/latest/src/ed25519_consensus/signing_key.rs.html#16-21

The verification key is no longer passed around.